google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/internal/RowSetUtil.java

```diff
@@ -0,0 +1,346 @@
+    // An empty RowSet represents a full table scan. Make that explicit so that there is RowRange to
+    // split.
+    if (RowSet.getDefaultInstance().equals(rowSet)) {
+      rowSet = RowSet.newBuilder().addRowRanges(RowRange.getDefaultInstance()).build();
+    }
+
+    // Create sorted copies of the ranges and keys in the RowSet
+    ByteString[] rowKeys =
+        rowSet.getRowKeysList().toArray(new ByteString[rowSet.getRowKeysCount()]);
+    RowRange[] rowRanges =
+        rowSet.getRowRangesList().toArray(new RowRange[rowSet.getRowRangesCount()]);
+
+    Arrays.sort(rowKeys, ByteStringComparator.INSTANCE);
+    Arrays.sort(rowRanges, RANGE_START_COMPARATOR);
+
+    List<RowSet> results = Lists.newArrayList();
+
+    // Track consumption of input ranges & keys.
+    int rowKeysStart = 0;
+    int rowRangesStart = 0;
+
+    // Keys and ranges that lie before the current split point,
+    RowSet.Builder segment = RowSet.newBuilder();
+    boolean isSegmentEmpty = true;
+
+    for (ByteString splitPoint : splitPoints) {
+      Preconditions.checkState(!splitPoint.isEmpty(), "Split point can't be empty");
+
+      // Consume all of the row keys that lie to the left of the split point. Consumption is
+      // designated by advancing rowKeysStart.
+      for (int i = rowKeysStart; i < rowKeys.length; i++) {
+        ByteString rowKey = rowKeys[i];
+        if (ByteStringComparator.INSTANCE.compare(rowKey, splitPoint) <= 0) {
+          segment.addRowKeys(rowKey);
+          isSegmentEmpty = false;
+          rowKeysStart++;
+        } else {
+          // This key and all following keys belong to a later segment.
+          break;
+        }
+      }
+
+      // Consume all of the ranges that lie before the split point (splitting the range if
+      // necessary). Consumption is designated by advancing rowRangesStart.
+      for (int i = rowRangesStart; i < rowRanges.length; i++) {
+        RowRange rowRange = rowRanges[i];
+
+        // Break early when encountering the first start point that is past the split point.
+        // (The split point is the inclusive end of of the segment)
+        int startCmp = StartPoint.extract(rowRange).compareTo(new StartPoint(splitPoint, true));
+        if (startCmp > 0) {
+          break;
+        }
+
+        // Some part of this range will be in the segment.
+        isSegmentEmpty = false;
+
+        // Figure out the endpoint and remainder.
+        int endCmp = EndPoint.extract(rowRange).compareTo(new EndPoint(splitPoint, true));
+        if (endCmp <= 0) {
+          // The range is fully contained in the segment.
+          segment.addRowRanges(rowRange);
+
+          // Consume the range, but take care to shift partially consumed ranges.
+          // Partially consumed ranges will only exist if the original RowSet had overlapping
+          // ranges, this should be a rare occurrence.
+          System.arraycopy(
```

Comment:
Pretty difficult to understand this. Maybe do some preprocessing to get rid of overlaps first? Preprocessing will make resulting ranges simpler.

Reply:
I played around with implementing range merging and it's non-trivial. It creates quite a bit of noise and has very little practical benefit because most queries will not have overlapping rows. At some point in the future, if we need to implement RowSet merging, then it would make sense to premerge the ranges. But for now I updated the comments to give a better explanation of whats happening

Reply:
sgtm